### PR TITLE
CNDIT-1671: Add Kafka listener for Notification changing event

### PR DESCRIPTION
**Pull request by sveselev**
## Description

Introduces a dedicated Kafka listener for handling Notification changes in the Investigation service; includes enhancements focused on performance and code readability.

- `InvestigationService.java`: added the new processNotification method and improved exception handling.
- `NotificationUpdate.java`: DTO to handle the `sp_notification_event` output.
- `020-sp_notification_event-002.sql`: updated version of the procedure containing `notification_uid` in the ouptput
- `UtilHelper.java`: moved to `common-util`

## JIRA
[CNDIT-1671](https://cdc-nbs.atlassian.net/browse/CNDIT-1671)

### CODE REVIEW

**Review by upasanapattnaik-eq — COMMENTED** [review]
In the event of a Notification being created at the same time as an investigation- PHC event and Notification event, we will be expecting two events, correct?

File: investigation-service/src/main/java/gov/cdc/etldatapipeline/investigation/service/InvestigationService.java [hunk 1]
```diff
@@ -76,66 +91,76 @@ public class InvestigationService {
...
-            logger.error(msg, e.getMessage());
-            throw new RuntimeException(e);
+            String msg = "Error processing Investigation data" +
+                    (!publicHealthCaseUid.isEmpty() ? " with ids '" + publicHealthCaseUid + "': " : ": " + e.getMessage());
+            throw new RuntimeException(msg, e);
         }
     }
 
-    private void processPhcFactDatamart(String publicHealthCaseUid) {
+    public void processNotification(String value) {
```
Comment (upasanapattnaik-eq) [inline investigation-service/src/main/java/gov/cdc/etldatapipeline/investigation/service/InvestigationService.java:145]: 👍

File: liquibase-service/src/main/resources/db/odse/routines/020-sp_notification_event-002.sql [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+                                          act_relationship act WITH (NOLOCK)
+                                              join public_health_case phc WITH (NOLOCK) on act.target_act_uid = phc.public_health_case_uid
+                                              join participation part with (nolock) ON part.type_cd='SubjOfPHC' AND part.act_uid=act.target_act_uid
+                                              join person per with (nolock) ON per.cd='PAT' AND per.person_uid = part.subject_entity_uid
+                                      WHERE
+                                          act.source_act_uid = notif.notification_uid
+                                        AND notif.cd not in ('EXP_NOTF', 'SHARE_NOTF', 'EXP_NOTF_PHDC','SHARE_NOTF_PHDC')
+                                        AND act.source_class_cd = 'NOTF'
+                                        AND act.target_class_cd = 'CASE' FOR json path,INCLUDE_NULL_VALUES
+                                  ) AS investigation_notifications
```
Comment (upasanapattnaik-eq) [inline liquibase-service/src/main/resources/db/odse/routines/020-sp_notification_event-002.sql:88]: Would it be better to rename this to investigation_notification?

**Review by upasanapattnaik-eq — APPROVED** [review]
LGTM! 💯

**Comment by sveselev** [comment]
> In the event of a Notification being created at the same time as an investigation- PHC event and Notification event, we will be expecting two events, correct?

Correct!

**Comment by upasanapattnaik-eq** [comment]
> > In the event of a Notification being created at the same time as an investigation- PHC event and Notification event, we will be expecting two events, correct?
> 
> Correct!

Sounds good. The data is the same at that point of time. Would there be any cases that we would expect differently?

**Pull request merged**